src/dune_engine/fs_memo.ml

```diff
@@ -234,10 +270,11 @@ let with_lexbuf_from_file path ~f =
 (* When a file or directory is created or deleted, we need to also invalidate
    the parent directory, so that the [dir_contents] queries are re-executed. *)
 let invalidate_path_and_its_parent path =
-  Memo.Invalidation.combine (invalidate_path path)
+  Memo.Invalidation.combine
+    (invalidate_path_both path)
```

Comment:
I think you are invalidating way too many things here. Could you add a comment clarifying why we invalidate each path?